src/Handler/Search.hs

```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
+        provideRep (jsonOutput results page hasMore)
   where
-    htmlOutput :: Text -> [SearchResult] -> Handler Html
-    htmlOutput query results = do
+    getPageLinkRenderer :: Handler (Maybe (Int -> Text))
+    getPageLinkRenderer = runMaybeT $ do
+      cr <- MaybeT getCurrentRoute
+      MaybeT $ do
+        urp <- getUrlRenderParams
+        getParams <- Map.fromList . reqGetParams <$> getRequest
+        return $ Just $ \p ->
+          urp cr $ Map.toList
+                 $ Map.insert "page" (pack $ show p) getParams
+
+    mkPaginationW :: Int -> Bool -> Handler (Maybe Widget)
+    mkPaginationW page hasMore = getPageLinkRenderer >>= \mr -> case mr of
+        Nothing -> return Nothing
+        Just mkPageLink ->
+          let allPages = [1..page]
+              mmNextPageLink = if hasMore
+                                  then Just $ if page >= maxPages
+                                    then Nothing
+                                    else Just $ mkPageLink $ page + 1
+                                  else Nothing
+              mPrevPageLink = if page > 1
+                                  then Just $ mkPageLink $ page - 1
+                                  else Nothing
+           in return $ if isNothing mmNextPageLink && isNothing mPrevPageLink
+                then Nothing
+                else Just $ $(widgetFile "pagination")
+
+    mkPaginationHeader :: Int -> Bool -> Handler (Maybe (Text, Text))
+    mkPaginationHeader page hasMore = (go =<<) <$> getPageLinkRenderer
+      where
+      go mkPageLink =
+          let mNextPageLink = if hasMore && page < maxPages
+                                  then Just $ mkPageLink $ page + 1
+                                  else Nothing
+              mPrevPageLink = if page > 1
+                                  then Just $ mkPageLink $ page - 1
+                                  else Nothing
+              links = catMaybes $ [ ("next", ) <$> mNextPageLink
+                                  , ("prev", ) <$> mPrevPageLink
+                                  ]
+           in if null links
+                then Nothing
+                else Just $ ("Link", T.intercalate ", " $ renderLink <$> links)
+        where renderLink :: (Text, Text) -> Text
+              renderLink (rel, url) = "<" <> url <> ">; rel=" <> (T.pack $ show rel)
+
+    htmlOutput :: Text -> [SearchResult] -> Int -> Bool -> Handler Html
+    htmlOutput query results page hasMore = do
       fr <- getFragmentRender
+      mPaginationW <- mkPaginationW page hasMore
       content <- defaultLayout $(widgetFile "search")
       sendResponseStatus ok200 content
 
-    jsonOutput = fmap toJSON . traverse searchResultToJSON
+    jsonOutput :: [SearchResult] -> Int -> Bool -> Handler Value
+    jsonOutput results page hasMore = do
+        when (hasMore && page <= maxPages) $ void $ do
+          mkPaginationHeader page hasMore >>= \mHeader -> case mHeader of
```

Comment:
I didn't want to add any syntax extensions, but I think `LambdaCase` would be pretty cool for all these cases:

```
mkPaginationHeader page hasMore >>= case
  Nothing -> pure ()
  Just ...
```